File: leetcode/questions/918.环形子数组的最大和.go
/*
 * @lc app=leetcode.cn id=918 lang=go
 *
 * [918] 环形子数组的最大和
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
 func maxSubarraySumCircular(nums []int) int {
	total, maxSum, minSum, currMax, currMin := nums[0], nums[0], nums[0], nums[0], nums[0]

	for i := 1; i < len(nums); i++ {
			total += nums[i]
			currMax = max(currMax+nums[i], nums[i])
			maxSum  = max(maxSum, currMax)
			currMin = min(currMin+nums[i], nums[i])
			minSum  = min(minSum, currMin)
	}

	//等价于if maxSum < 0
	if total == minSum  {
			return maxSum
	} else {
			return max(maxSum, total - minSum)
	}
}

func max(a, b int) int {
	if a > b {
			return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
			return a
	}
	return b
}
// @lc code=end

File: code/questions/918.环形子数组的最大和.js
/*
 * @lc app=leetcode.cn id=918 lang=javascript
 *
 * [918] 环形子数组的最大和
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
const getMaxArr = (nums) => {
  let ans = nums[0], sum = 0;
  for (let i = 0, leng = nums.length;i < leng;  i++){
    if (sum > 0) {
      sum += nums[i];
    } else {
      sum = nums[i];
    }
    ans = ans > sum ? ans : sum;
  }
  return ans;
}
const getMinArr = (nums) => {
  let ans = nums[0], sum = 0;
  for (let i = 0, leng = nums.length;i < leng;  i++){
    if (sum < 0) {
      sum += nums[i];
    } else {
      sum = nums[i];
    }
    ans = ans < sum ? ans : sum;
  }
  return ans;
}
var maxSubarraySumCircular = function (nums) {
  // const sum = nums.reduce((pre, cur) => pre + cur);
  let sum = 0, positiveCount = 0, negativeCount = 0, minVal = nums[0], maxVal = nums[0];
  for (let i = 0, leng = nums.length;i < leng;i++){
    sum += nums[i];
    if (nums[i] > 0) {
      positiveCount++;
      minVal = minVal < nums[i] ? minVal : nums[i];
    }
    if (nums[i] < 0) {
      negativeCount++;
      maxVal = maxVal > nums[i] ? maxVal : nums[i];
    }
  }
  if (positiveCount === nums.length) {
    return sum;
  }
  if (negativeCount === nums.length) {
    console.log(negativeCount, maxVal)
    return maxVal;
  }

  const numsMin = getMinArr(nums);
  const numsMax = getMaxArr(nums);
  return numsMax > sum - numsMin ? numsMax : sum - numsMin;
};
// @lc code=end

